### PR TITLE
Value Converters for All Core Property Editors

### PR DESCRIPTION
**Pull request by Jeavon**
Finally it's here! I'm opening this PR for discussion/review/revision etc... This is the combined work of James Coxhead, Lars-Erik Aabech and also the port of my own value converters package into core (which includes the work of many others over the years!).

The objective:
Make Razor super simple for devs and of course enable Models Builder to generate classes with the correct Types

So whats in here?
Value converters for the following property editors:
- CheckboxList
- ColorPicker
- DropDownListMultiple
- DropDownListMultipleWithKeys
- DropDownList
- DropdownListWithKeys
- EmailAddress
- MemberGroup
- RadioButtonList
- Silder
- Tags
- Upload
- UserPicker
- RelatedLinks
- ContentPicker
- MediaPicker (legacy - disabled as requested by @zpqrtbnk as support TBC)
- MemberPicker
- MultinodeTreePicker
- MultipleMediaPicker

A new umbracoSettings.config setting (EnablePropertyValueConverters) to enable all the above converters, without this setting everything should be exactly as it is now. As previously discussed with @Shazwazza & @zpqrtbnk back in June, these converters do not support dynamics so this setting cannot be enabled if dynamic published content is being used in Razor!
As there was an existing RelatedLinks Value Converter, this is now enabled when EnablePropertyValueConverters is false.

Where possible TypeConverters have been added to allow for simpler upgrade paths/flexibility, e.g Content Picker will still work when a dev has already implemented `Model.Content.GetPropertyValue<int>("contentPicker")`

Outstanding questions:
- Should DropDownListWithKeys and DropDownListMultipleWithKeys return Umbraco.Core.Models.PreValue?
- Should/Could Tags return collections of Umbraco.Web.Models.TagModel?
- The Content and MultiNode Picker converters currently exclude conversion of the InternalRedirect & Redirect property aliases, instead the usage of these internally could be made to support IPublishedContent? e.g. https://github.com/umbraco/Umbraco-CMS/compare/dev-v7.6...temp-u4-7318?expand=1#diff-5e84c87cc44ee610f96d418da8c386d9R32

Road blocks:
Several of the converters (MultipleMediaPicker, Tags & Slider) require the usage of prevalues, in my package I have a implemented my own caching (https://github.com/Jeavon/Umbraco-Core-Property-Value-Converters/blob/v3/Our.Umbraco.PropertyConverters/MultipleMediaPickerPropertyConverter.cs#L213) to ensure performance but I have removed this for Core as discussed with @zpqrtbnk but we will need a mechanism for caching these prevalues before this PR becomes fully viable. I have created a issue for this work item http://issues.umbraco.org/issue/U4-8862

I'm sure there's loads more I've missed but I'm sure it will come up....

### CODE REVIEW

**Comment by Shazwazza** [comment]
Just FYI: I'm going to pull this in but will also be adding review notes - these are my own notes and I will fix these things up in the next week or two after this is pulled in so there's nothing anyone needs to do specifically.

File: src/Umbraco.Core/EnumerableExtensions.cs [hunk 1]
```diff
@@ -301,5 +301,21 @@ public static bool UnsortedSequenceEqual<T>(this IEnumerable<T> source, IEnumera
...
+        ///<summary>
+        /// Returns the items of the given enumerable as a pure enumerable.
+        /// <remarks>
+        /// When quering lists using methods such as <see cref="M:List.Where"/>, the result, despite appearing to look like and quack like an 
+        /// <see cref="T:Enumerable{T}"/> the type is actually an instance of <see cref="T:System.Linq.Enumerable.WhereEnumerableIterator"/>
+        /// </remarks>
+        /// </summary>
+        ///<param name="source">The item to find.</param>
+        ///<returns>The index of the first matching item, or -1 if the item was not found.</returns>
+        internal static IEnumerable<T> Yield<T>(this IEnumerable<T> source)
```
Comment (Shazwazza) [inline src/Umbraco.Core/EnumerableExtensions.cs]: This needs a thorough review it seems highly unnecessary

File: src/Umbraco.Core/Models/PublishedContent/PublishedContentEnumerable.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{
+    /// <summary>
+    /// The published content enumerable, this model is to allow ToString to be overriden for value converters to support legacy requests for string values
+    /// </summary>
+    public class PublishedContentEnumerable : IEnumerable<IPublishedContent>
+    {
+        /// <summary>
+        /// The items in the collection
+        /// </summary>
+        private readonly List<IPublishedContent> _items = new List<IPublishedContent>();
```
Comment (Shazwazza) [inline src/Umbraco.Core/Models/PublishedContent/PublishedContentEnumerable.cs]: Why is this a list, why not just store the original IEnumerable?

File: src/Umbraco.Core/PropertyEditors/ValueConverters/DropdownListValueConverter.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco.Core.PropertyEditors.ValueConverters
+{
+    public class DropdownListValueConverter : PropertyValueConverterBase, IPropertyValueConverterMeta
+    {
+        public override bool IsConverter(PublishedPropertyType propertyType)
+        {
+            return propertyType.PropertyEditorAlias.InvariantEquals(Constants.PropertyEditors.DropDownListAlias);
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/DropdownListValueConverter.cs]: A few of these are missing the if (UmbracoConfig.For.UmbracoSettings().Content.EnablePropertyValueConverters) check

File: src/Umbraco.Core/PropertyEditors/ValueConverters/EmailAddressValueConverter.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco.Core.PropertyEditors.ValueConverters
+{
+    public class EmailAddressValueConverter : PropertyValueConverterBase, IPropertyValueConverterMeta
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/EmailAddressValueConverter.cs]: What is the point of this converter?

File: src/Umbraco.Core/PropertyEditors/ValueConverters/MemberGroupPickerValueConverter.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco.Core.PropertyEditors.ValueConverters
+{
+    public class MemberGroupPickerValueConverter : PropertyValueConverterBase, IPropertyValueConverterMeta
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/MemberGroupPickerValueConverter.cs]: What is the purpose of this converter?

File: src/Umbraco.Core/PropertyEditors/ValueConverters/TagsValueConverter.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        /// <param name="dataTypeId">
+        /// The data type id.
+        /// </param>
+        /// <returns>
+        /// The <see cref="bool"/>.
+        /// </returns>
+        private static bool JsonStorageType(int dataTypeId)
+        {
+            // ** This must be cached (U4-8862) **
+            var dts = ApplicationContext.Current.Services.DataTypeService;
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/TagsValueConverter.cs]: These services should be ctor injected so they can be tested

File: src/Umbraco.Core/PropertyEditors/ValueConverters/SliderValueConverter.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        /// <param name="dataTypeId">
+        /// The data type id.
+        /// </param>
+        /// <returns>
+        /// The <see cref="bool"/>.
+        /// </returns>
+        private bool IsRangeDataType(int dataTypeId)
+        {
+            // ** This must be cached (U4-8862) **
+            var dts = ApplicationContext.Current.Services.DataTypeService;
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/SliderValueConverter.cs]: These services should be ctor injected so they can be tested

File: src/Umbraco.Core/PropertyEditors/ValueConverters/ColorPickerValueConverter.cs [hunk 8]
```diff
@@ -1,18 +1,34 @@
-﻿using Umbraco.Core.Models.PublishedContent;
+﻿using System;
+using Umbraco.Core.Configuration;
+using Umbraco.Core.Models.PublishedContent;
 
 namespace Umbraco.Core.PropertyEditors.ValueConverters
 {
-    public class ColorPickerValueConverter : PropertyValueConverterBase
+    public class ColorPickerValueConverter : PropertyValueConverterBase, IPropertyValueConverterMeta
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/ColorPickerValueConverter.cs]: Most of these simple converters shouldn't be using IPropertyValueConverterMeta and use attributes instead.

File: src/Umbraco.Core/PropertyEditors/ValueConverters/CheckboxListValueConverter.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Umbraco.Core.Configuration;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco.Core.PropertyEditors.ValueConverters
+{
+    public class CheckboxListValueConverter : PropertyValueConverterBase, IPropertyValueConverterMeta
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/CheckboxListValueConverter.cs]: Since these will all be default property value converters, they need the DefaultPropertyValueConverterAttribute attribute

File: src/Umbraco.Core/PropertyEditors/ValueConverters/ColorPickerValueConverter.cs [hunk 10]
```diff
@@ -1,18 +1,34 @@
...
+	        {
+				return propertyType.PropertyEditorAlias.InvariantEquals(Constants.PropertyEditors.ColorPickerAlias);
+			}
+	        return false;
         }
 
         public override object ConvertDataToSource(PublishedPropertyType propertyType, object source, bool preview)
         {
             // make sure it's a string
             return source.ToString();
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/ColorPickerValueConverter.cs]: missing null check

File: src/Umbraco.Core/PropertyEditors/ValueConverters/UserPickerValueConverter.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                return user;
+            }
+
+            return userId;
+        }
+
+        public Type GetPropertyValueType(PublishedPropertyType propertyType)
+        {
+            return typeof(IUser);
```
Comment (Shazwazza) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/UserPickerValueConverter.cs]: This cannot be accepted, under no circumstances are we allowing returning business logic classes to be accessible on the front-end, especially an IUser object which contains sensitive information.

File: src/Umbraco.Core/PropertyEditors/ValueConverters/EmailAddressValueConverter.cs (hunk 4 above)
Comment (Jeavon) [inline src/Umbraco.Core/PropertyEditors/ValueConverters/EmailAddressValueConverter.cs]: Purely so that Models Builder generates class properties of type string instead of object

File: src/Umbraco.Web/PropertyEditors/ValueConverters/RelatedLinksEditorValueConvertor.cs [hunk 12]
```diff
@@ -1,106 +1,124 @@
...
-                        }    
-                    }
-                    return obj;
+                    relatedLinks.Add(relatedLink);
                 }
-                catch (Exception ex)
+                else
                 {
-                    LogHelper.Error<RelatedLinksEditorValueConvertor>("Could not parse the string " + sourceString + " to a json object", ex);
+                    LogHelper.Warn<RelatedLinks>(
```
Comment (Shazwazza) [inline src/Umbraco.Web/PropertyEditors/ValueConverters/RelatedLinksEditorValueConvertor.cs]: Logging for RelatedLinks when it's actually for RelatedLinksPropertyConverter

File: src/Umbraco.Web/RelatedLinksTypeConverter.cs [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+using Umbraco.Core;
+using Umbraco.Web.Models;
+
+namespace Umbraco.Web
+{
+    public class RelatedLinksTypeConverter : TypeConverter
+    {
+        public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType)
+        {
+            var convertableTypes = new[]
```
Comment (Shazwazza) [inline src/Umbraco.Web/RelatedLinksTypeConverter.cs]: This should be a static property, you are allocating a new array everytime this executes.

File: src/Umbraco.Web/Models/RelatedLinks.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        /// </summary>
+        internal string PropertyData
+        {
+            get
+            {
+                return this._propertyData;
+            }
+        }
+
+        public bool Any()
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/RelatedLinks.cs]: Why is this needed? This class inherits from IEnumerable<T>

File: src/Umbraco.Web/Models/RelatedLinkBase.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+// <summary>
+//   Defines the RelatedLink base type.
+// </summary>
+// --------------------------------------------------------------------------------------------------------------------
+
+using Newtonsoft.Json;
+
+namespace Umbraco.Web.Models
+{
+    public class RelatedLinkBase
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/RelatedLinkBase.cs]: Should be abstract

File: src/Umbraco.Web/Models/RelatedLinkData.cs [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+// <summary>
+//   Defines the RelatedLink json data type.
+// </summary>
+// --------------------------------------------------------------------------------------------------------------------
+
+using Newtonsoft.Json;
+
+namespace Umbraco.Web.Models
+{
+    public class RelatedLinkData : RelatedLinkBase
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/RelatedLinkData.cs]: Should be internal, only used internally
